CODETOOLS-7902853: JMH: Fix lint and deprecation warnings

nielsbasjes opened the pull request:
In my projects I always use -Werror as the compiler flag as one of the means to get a high standard for code quality.
In this project https://github.com/nielsbasjes/yauaa I have a module that does the benchmarks.
When I update the JMH from 1.26 to 1.28 my build fails with 
```
[WARNING] COMPILATION WARNING : 
[INFO] -------------------------------------------------------------
[WARNING] /home/nbasjes/workspace/Prive/yauaa_/benchmarks/target/generated-sources/annotations/nl/basjes/parse/useragent/benchmarks/jmh_generated/AnalyzerBenchmarks_android6Chrome46_jmhTest.java:[324,67] redundant cast to long
[WARNING] /home/nbasjes/workspace/Prive/yauaa_/benchmarks/target/generated-sources/annotations/nl/basjes/parse/useragent/benchmarks/jmh_generated/AnalyzerBenchmarks_android6Chrome46_jmhTest.java:[324,83] redundant cast to long
[WARNING] /home/nbasjes/workspace/Prive/yauaa_/benchmarks/target/generated-sources/annotations/nl/basjes/parse/useragent/benchmarks/jmh_generated/AnalyzerBenchmarks_androidPhone_jmhTest.java:[324,67] redundant cast to long
[WARNING] /home/nbasjes/workspace/Prive/yauaa_/benchmarks/target/generated-sources/annotations/nl/basjes/parse/useragent/benchmarks/jmh_generated/AnalyzerBenchmarks_androidPhone_jmhTest.java:[324,83] redundant cast to long
```
and at the end it fails
```
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/nbasjes/workspace/Prive/yauaa_/benchmarks/target/generated-sources/annotations/nl/basjes/parse/useragent/benchmarks/jmh_generated/AnalyzerBenchmarks_android6Chrome46_jmhTest.java: warnings found and -Werror specified
[INFO] 1 error
```

Turns out this problem is caused by a change in code generated by JMH.

This pull request fixes this problem.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Issue
 * [CODETOOLS-7902853](https://bugs.openjdk.java.net/browse/CODETOOLS-7902853): JMH: Fix lint and deprecation warnings


### Reviewers
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - Committer)


### Contributors
 * Aleksey Shipilev `<shade@openjdk.org>`

### Download
To checkout this PR locally:
`$ git fetch https://git.openjdk.java.net/jmh pull/31/head:pull/31`
`$ git checkout pull/31`

To update a local copy of the PR:
`$ git checkout pull/31`
`$ git pull https://git.openjdk.java.net/jmh pull/31/head`

shipilev reviewed (changes requested):
Right. There are also three other places where these casts happen, grep for `new BenchmarkTaskResult` in the same file. I would also need to check how to regression-test this...

nielsbasjes commented:
Yes, I see them.
I'll try to see what happens if I simply add the -Werror to the entire build, that is essentially what happens in my project.
Note: I have submitted the OCA this morning; apparently this takes time.

nielsbasjes commented:
I did some checks and tests.

I reverted my change (locally) and added `-Werror` to the compiler on the `jmh-core-it` which did not cause the same fail I have.

My best guess at this moment is that it is related to my project using Java 8 and 11, and this project explicitly uses java 1.7.

Also I get a (correct) warning about the usage of `Unsafe` (used to cause a segmentation fault in a test) which I was unsuccessful in suppressing.

Also in all of the three other situations the input field is a `double` that is then cast to a `long`, which seems fine to me.

shipilev commented:
> My best guess at this moment is that it is related to my project using Java 8 and 11, and this project explicitly uses java 1.7.

I think it also requires `-Xlint` to be enabled. See this patch, which hooks up `-Xlint` and `-Werror` to JMH compilation tests:
  https://cr.openjdk.java.net/~shade/jmh/Werror-1.patch 

Please mix it in to this PR. (Then maybe say `/contributor add @shipilev` here).

> Also in all of the three other situations the input field is a `double` that is then cast to a `long`, which seems fine to me.

Yes!

nielsbasjes commented:
I'll pull that in.
And cleanup some other things I ran into (in separate commits).

nielsbasjes commented:
/contributor add @shipilev

shipilev commented on jmh-core/src/test/java/org/openjdk/jmh/util/TestListStatistics.java:
```diff
@@ -240,17 +240,17 @@ public strictfp void testEmpty() {
-        Assert.assertEquals(Double.NaN, s.getConfidenceIntervalAt(0.50)[1]);
-        Assert.assertEquals(Double.NaN, s.getPercentile(0));
-        Assert.assertEquals(Double.NaN, s.getPercentile(100));
+        Assert.assertEquals(Double.NaN, s.getSum(), 0.0000001);
+        Assert.assertEquals(Double.NaN, s.getMin(), 0.0000001);
+        Assert.assertEquals(Double.NaN, s.getMax(), 0.0000001);
+        Assert.assertEquals(Double.NaN, s.getMean(), 0.0000001);
+        Assert.assertEquals(Double.NaN, s.getMeanErrorAt(0.5), 0.0000001);
+        Assert.assertEquals(Double.NaN, s.getVariance(), 0.0000001);
+        Assert.assertEquals(Double.NaN, s.getStandardDeviation(), 0.0000001);
+        Assert.assertEquals(Double.NaN, s.getConfidenceIntervalAt(0.50)[0], 0.0000001);
+        Assert.assertEquals(Double.NaN, s.getConfidenceIntervalAt(0.50)[1], 0.0000001);
+        Assert.assertEquals(Double.NaN, s.getPercentile(0), 0.0000001);
+        Assert.assertEquals(Double.NaN, s.getPercentile(100), 0.0000001);
```
Please introduce the `static final double ACCURACY = 0.0000001`, and use it?

nielsbasjes replied:
Done in all places.

shipilev commented on jmh-core/src/test/java/org/openjdk/jmh/util/TestListStatistics.java:
```diff
@@ -259,17 +259,17 @@ public strictfp void testSingle() {
-        Assert.assertEquals(Double.NaN, s.getConfidenceIntervalAt(0.50)[1]);
-        Assert.assertEquals(42.0D, s.getPercentile(0));
-        Assert.assertEquals(42.0D, s.getPercentile(100));
+        Assert.assertEquals(42.0D, s.getSum(), 0.0000001);
+        Assert.assertEquals(42.0D, s.getMin(), 0.0000001);
+        Assert.assertEquals(42.0D, s.getMax(), 0.0000001);
+        Assert.assertEquals(42.0D, s.getMean(), 0.0000001);
+        Assert.assertEquals(Double.NaN, s.getMeanErrorAt(0.5), 0.0000001);
+        Assert.assertEquals(Double.NaN, s.getVariance(), 0.0000001);
+        Assert.assertEquals(Double.NaN, s.getStandardDeviation(), 0.0000001);
+        Assert.assertEquals(Double.NaN, s.getConfidenceIntervalAt(0.50)[0], 0.0000001);
+        Assert.assertEquals(Double.NaN, s.getConfidenceIntervalAt(0.50)[1], 0.0000001);
+        Assert.assertEquals(42.0D, s.getPercentile(0), 0.0000001);
+        Assert.assertEquals(42.0D, s.getPercentile(100), 0.0000001);
```
Ditto, `ACCURACY` constant.

shipilev commented on jmh-core/src/test/java/org/openjdk/jmh/util/TestListStatistics.java:
```diff
@@ -405,8 +405,8 @@ public strictfp void testRawDataIterator() {
             Map.Entry<Double, Long> entry = listIter.next();
-            Assert.assertEquals(entry.getKey(), item);
-            Assert.assertEquals(entry.getValue().longValue(), 1L);
+            Assert.assertEquals(entry.getKey(), item, 0.0000001);
```
Ditto, `ACCURACY` constant.

shipilev commented on jmh-core/src/test/java/org/openjdk/jmh/util/TestMultisetStatistics.java:
```diff
@@ -234,17 +234,17 @@ public strictfp void testEmpty() {
-        Assert.assertEquals(Double.NaN, s.getConfidenceIntervalAt(0.50)[1]);
-        Assert.assertEquals(Double.NaN, s.getPercentile(0));
-        Assert.assertEquals(Double.NaN, s.getPercentile(100));
+        Assert.assertEquals(Double.NaN, s.getSum(), 0.0000001);
```
Ditto, `ACCURACY`.

shipilev reviewed (changes requested):
Thanks, this looks good. I edited this PR to get it hooked to bug systems. Please introduce `ACCURACY` constants, and then we are good to go patch-wise. We would need to wait for OCA to clear for integration, though.

shipilev reviewed (approved):
Looks good. Once you have OCA approved, bot would lift `oca` tag, and then instruct you to say `/integrate`. We can ping @robilad to see if there are problems with OCA.

nielsbasjes commented:
It has been 'open' for several days. 
So I resubmitted the OCA just now.
@robilad It it good now?

nielsbasjes commented:
/integrate

shipilev commented:
/sponsor

shipilev commented:
Delivered in JMH 1.29.
